clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -6193,7 +6193,13 @@ static void processSYCLKernel(Sema &S, FunctionDecl *FD, MangleContext &MC) {
   if (S.LangOpts.SYCLIsDevice) {
     S.ConstructOpenCLKernel(FD, MC);
   } else if (S.LangOpts.SYCLIsHost) {
-    CXXRecordDecl *CRD = (*FD->param_begin())->getType()->getAsCXXRecordDecl();
+    QualType KernelParamTy = (*FD->param_begin())->getType();
+    const CXXRecordDecl *CRD;
+    if (KernelParamTy->isReferenceType())
+      CRD =
+          const_cast<CXXRecordDecl *>(KernelParamTy->getPointeeCXXRecordDecl());
```

Comment:
The problem still exists here. It  is still missing generating subgroup attribute on host for AST and IR. So all tests associated with this PR will fail.

Reply:
The problem seems happening with this test case:

class Functor16 {
public:
  [[intel::reqd_sub_group_size(4)]] void operator()() const {
  }
};

template <typename name, typename Func>
__attribute__((sycl_kernel)) void kernel(const Func &kernelFunc) {
  kernelFunc();
}

void bar() {
  Functor16 f16;
  kernel<class kernel_name1>(f16);

}